File: src/shared/inventory.ts
import type { AgentId } from './transcript';

// What the settings page shows per agent: where its executable is, what its own config files declare (MCP servers / skills / rules).
// Read-only: Acpira lists and opens these files, it never writes them

export type McpTransport = 'stdio' | 'http' | 'sse';

export type InventoryScope = 'user' | 'project';

export interface InventoryFile {
  path: string;
  scope: InventoryScope;
  exists: boolean;
  // Bytes; only when it exists
  size?: number;
}

export interface InventoryMcp {
  name: string;
  transport: McpTransport;
  // Command line (stdio) or URL (http / sse), for display
  target: string;
  // The config file it was declared in
  source: string;
  scope: InventoryScope;
  enabled: boolean;
}

export interface InventorySkill {
  name: string;
  description?: string;
  // The SKILL.md
  path: string;
  scope: InventoryScope;
}

// Runtime facts known only from a live process's initialize response; absent until a session of that agent has been opened
export interface AgentRuntimeInfo {
  name?: string;
  version?: string;
  mcp?: { http: boolean; sse: boolean };
}

// Versions read off an npm-installed ACP adapter (AgentDef.adapter) and the runtime it bundles. Read from package.json
// files on disk — the CLI is never invoked. Missing pieces stay undefined; a non-adapter agent has no adapter field at all
export interface AdapterInfo {
  adapter?: { name: string; version?: string; root?: string };
  // `override`/`overrideEnv` when the env var redirects the bundled runtime to a different binary
  engine?: { name: string; version?: string; override?: string; overrideEnv?: string };
}

// How far launching the agent got last time: the executable was found (binary), the process spawned, the ACP handshake
// answered, and session/new succeeded or asked for sign-in. Recorded by the controls probe and by real session starts
export type AgentHealthStage = 'spawn_failed' | 'handshake_failed' | 'auth_required' | 'ready';
export interface AgentHealth {
  stage: AgentHealthStage;
  at: string;
  error?: string;
  source: 'probe' | 'session';
}

export interface AgentInventory {
  agent: AgentId;
  // Resolved executable; null when not found
  binary: string | null;
  runtime?: AgentRuntimeInfo;
  adapter?: AdapterInfo;
  health?: AgentHealth;
  // Whether a second session/prompt mid-turn steers the running turn (registry knowledge; 1.0 always queues)
  steer: boolean;
  config: InventoryFile[];
  mcp: InventoryMcp[];
  skills: InventorySkill[];
  rules: InventoryFile[];
  scannedAt: string;
}

File: rust/crates/acpira-shared/src/inventory.rs
//! Agent inventory shown on the settings page (mirror of src/shared/inventory.ts)

use serde::{Deserialize, Serialize};

use crate::transcript::AgentId;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum McpTransport {
  Stdio,
  Http,
  Sse,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InventoryScope {
  User,
  Project,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InventoryFile {
  pub path: String,
  pub scope: InventoryScope,
  pub exists: bool,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub size: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InventoryMcp {
  pub name: String,
  pub transport: McpTransport,
  pub target: String,
  pub source: String,
  pub scope: InventoryScope,
  pub enabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InventorySkill {
  pub name: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub description: Option<String>,
  pub path: String,
  pub scope: InventoryScope,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct McpCaps {
  pub http: bool,
  pub sse: bool,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentRuntimeInfo {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub name: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub version: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub mcp: Option<McpCaps>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AdapterPart {
  pub name: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub version: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub root: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnginePart {
  pub name: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub version: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub r#override: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub override_env: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct AdapterInfo {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub adapter: Option<AdapterPart>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub engine: Option<EnginePart>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentHealthStage {
  SpawnFailed,
  HandshakeFailed,
  AuthRequired,
  Ready,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HealthSource {
  Probe,
  Session,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentHealth {
  pub stage: AgentHealthStage,
  pub at: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub error: Option<String>,
  pub source: HealthSource,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentInventory {
  pub agent: AgentId,
  pub binary: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub runtime: Option<AgentRuntimeInfo>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub adapter: Option<AdapterInfo>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub health: Option<AgentHealth>,
  pub steer: bool,
  pub config: Vec<InventoryFile>,
  pub mcp: Vec<InventoryMcp>,
  pub skills: Vec<InventorySkill>,
  pub rules: Vec<InventoryFile>,
  pub scanned_at: String,
}
